File: backend/app/core/barcode/prn_generator.py
# -*- coding: utf-8 -*-
"""
Project      : SMRITI Retail OS
Author       : Jawahar Ramkripal Mallah
Designation  : Chief Systems Architect & Creator
Email        : support@smritibooks.com
Websites     : smritisys.com | smritibooks.com | erpnbook.com | aitdl.com
Version      : 3.40.0
Copyright    : © SMRITIBooks.com. All Rights Reserved.
License      : Proprietary Commercial Software

PRN Script Generator Engine
============================
Template-driven raw ZPL/TSPL PRN content generator for thermal barcode label printers.
Handles central token substitution via token_registry.build_token_dict() and built-in fallback scripts.
"""

import re
from typing import List, Dict, Any, Union, Optional
from .token_registry import build_token_dict


def safe_template_substitute(template: str, token_dict: Dict[str, str]) -> str:
    """
    Replaces `{token}` placeholders in raw ZPL/TSPL templates with token_dict values.
    Leaves any other literal `{` or `}` untouched instead of crashing.
    """
    if not template:
        return ""

    if not token_dict:
        return template

    pattern = r"\{(" + "|".join(re.escape(k) for k in token_dict.keys()) + r")\}"

    def _replace(match):
        key = match.group(1)
        return str(token_dict.get(key, match.group(0)))

    return re.sub(pattern, _replace, template)


# ---------------------------------------------------------------------------
# BUILT-IN FALLBACK TEMPLATES (ZPL & TSPL)
# ---------------------------------------------------------------------------

FALLBACK_TEMPLATES = {
    "ZPL": {
        "50x25": """^XA
^PW400
^LL200
^FO10,10^A0N,20,20^FD{brand}^FS
^FO10,35^A0N,18,18^FD{item_name}^FS
^FO10,55^A0N,16,16^FDStyle: {style}^FS
^FO10,72^A0N,16,16^FDSize: {size}  Color: {color}^FS
^FO10,95^BY2,3,40^BCN,40,Y,N,N^FD{barcode}^FS
^FO10,150^A0N,20,20^FDMRP: Rs.{mrp}.00^FS
^FO10,175^A0N,14,14^FDPkd: {pkd_date}  {company}^FS
^XZ""",
        "38x25": """^XA
^PW304
^LL200
^FO10,10^A0N,18,18^FD{brand}^FS
^FO10,30^A0N,16,16^FD{item_name}^FS
^FO10,48^A0N,14,14^FDSz:{size} Col:{color}^FS
^FO10,68^BY2,2,35^BCN,35,Y,N,N^FD{barcode}^FS
^FO10,118^A0N,18,18^FDMRP: Rs.{mrp}^FS
^XZ""",
        "100x50": """^XA
^PW800
^LL400
^FO20,20^A0N,32,32^FD{brand}^FS
^FO20,60^A0N,26,26^FDItem: {item_name}^FS
^FO20,95^A0N,22,22^FDStyle: {style}  |  Art: {article_no}^FS
^FO20,125^A0N,22,22^FDSize: {size}  |  Color: {color}  |  Mat: {material}^FS
^FO20,160^BY3,3,70^BCN,70,Y,N,N^FD{barcode}^FS
^FO20,250^A0N,36,36^FDMRP: Rs.{mrp}.00 (Incl. of all taxes)^FS
^FO20,295^A0N,20,20^FDPacked Date: {pkd_date}  |  HSN: {hsn}^FS
^FO20,325^A0N,18,18^FDMfd & Mkt by: {company}^FS
^XZ"""
    },
    "TSPL": {
        "50x25": """SIZE 50 mm, 25 mm
GAP 2 mm, 0 mm
DIRECTION 1
CLS
TEXT 10,10,"2",0,1,1,"{brand}"
TEXT 10,30,"1",0,1,1,"{item_name}"
TEXT 10,48,"1",0,1,1,"Style: {style} Sz:{size}"
BARCODE 10,70,"128",40,1,0,2,2,"{barcode}"
TEXT 10,120,"2",0,1,1,"MRP: Rs.{mrp}.00"
PRINT 1,1
""",
        "38x25": """SIZE 38 mm, 25 mm
GAP 2 mm, 0 mm
DIRECTION 1
CLS
TEXT 10,10,"2",0,1,1,"{brand}"
TEXT 10,30,"1",0,1,1,"{item_name}"
BARCODE 10,50,"128",35,1,0,2,2,"{barcode}"
TEXT 10,95,"2",0,1,1,"MRP: Rs.{mrp}"
PRINT 1,1
"""
    }
}


def generate_prn_script(
    items: List[Dict[str, Any]],
    raw_template: Optional[str] = None,
    protocol: str = "ZPL",
    label_size: str = "50x25",
    company_name: str = "SMRITI RETAIL"
) -> Dict[str, Any]:
    """
    Generates combined raw PRN script (ZPL/TSPL) for a batch of items.

    Args:
        items: List of item data dicts.
        raw_template: Custom ZPL/TSPL raw template string (optional).
        protocol: Printer command language ("ZPL", "TSPL", "EPL", "ESC/POS").
        label_size: Label dimensions profile ("50x25", "38x25", "100x50").
        company_name: Company legal name for footer token.

    Returns:
        dict: {
            "prn": str,
            "total_labels": int,
            "items_processed": int,
            "protocol": str
        }
    """
    if not items:
        return {
            "prn": "",
            "total_labels": 0,
            "items_processed": 0,
            "protocol": protocol
        }

    lang = (protocol or "ZPL").upper()
    size_key = (label_size or "50x25").lower()

    # Determine template to use
    template_to_use = raw_template
    if not template_to_use:
        lang_templates = FALLBACK_TEMPLATES.get(lang) or FALLBACK_TEMPLATES["ZPL"]
        template_to_use = lang_templates.get(size_key) or lang_templates.get("50x25") or FALLBACK_TEMPLATES["ZPL"]["50x25"]

    prn_chunks = []
    total_labels = 0

    for item in items:
        qty = int(item.get("print_qty") or item.get("stock_qty") or item.get("qty") or 1)
        if qty < 1:
            qty = 1

        token_dict = build_token_dict(item, company_name=company_name)
        single_label_script = safe_template_substitute(template_to_use, token_dict)

        # Repeat label output according to quantity
        for _ in range(qty):
            prn_chunks.append(single_label_script)
            total_labels += 1

    combined_prn = "\n".join(prn_chunks)

    return {
        "prn": combined_prn,
        "total_labels": total_labels,
        "items_processed": len(items),
        "protocol": lang
    }

File: src/services/prnGenerator.ts
/**
 * Project      : SMRITI Retail OS
 * Organization : SmritiSys
 * Author       : Jawahar Ramkripal Mallah
 * Designation  : Chief Systems Architect & Creator
 * Email        : support@smritibooks.com
 * Websites     : smritisys.com | smritibooks.com | erpnbook.com | aitdl.com
 * Version      : 5.3.0
 * Created      : 2026-07-27
 * Copyright    : © SmritiSys. All Rights Reserved.
 * License      : Proprietary Commercial Software
 */

import { Product } from "../types.js";

export interface PRNOptions {
  language: "TSPL" | "ZPL";
  widthMm?: number;
  heightMm?: number;
  gapMm?: number;
}

/**
 * Generates TSPL PRN script for TSC/TTP/GPRINTER thermal printers.
 */
export function generateTSPLScript(
  item: Product, 
  copies: number = 1, 
  options: { widthMm?: number; heightMm?: number; gapMm?: number } = {}
): string {
  const width = options.widthMm || 50;
  const height = options.heightMm || 25;
  const gap = options.gapMm || 2;
  const priceStr = `₹${Number(item.price || item.mrp || 0).toFixed(2)}`;
  const mrpStr = item.mrp ? `MRP: ₹${Number(item.mrp).toFixed(2)}` : "";
  const nameStr = (item.name || "Item").slice(0, 24).replace(/"/g, "'");
  const codeStr = (item.code || "").slice(0, 16).replace(/"/g, "'");
  const barcodeStr = item.barcode || item.code || "123456789";

  return `SIZE ${width} mm, ${height} mm
GAP ${gap} mm, 0 mm
DIRECTION 1,0
REFERENCE 0,0
OFFSET 0 mm
SET PEEL OFF
SET CUTTER OFF
CLS
TEXT 20,15,"3",0,1,1,"${nameStr}"
TEXT 20,40,"2",0,1,1,"SKU: ${codeStr} ${mrpStr ? "| " + mrpStr : ""}"
BARCODE 20,65,"128",40,1,0,2,2,"${barcodeStr}"
TEXT 20,115,"3",0,1,1,"OUR PRICE: ${priceStr}"
PRINT ${Math.max(1, copies)},1`;
}

/**
 * Generates ZPL PRN script for Zebra ZD421 / GX430 / GK420 thermal printers.
 */
export function generateZPLScript(
  item: Product, 
  copies: number = 1, 
  options: { widthMm?: number; heightMm?: number } = {}
): string {
  const priceStr = `INR ${Number(item.price || item.mrp || 0).toFixed(2)}`;
  const mrpStr = item.mrp ? `MRP:${Number(item.mrp).toFixed(2)}` : "";
  const nameStr = (item.name || "Item").slice(0, 24).replace(/[\^\~]/g, "");
  const codeStr = (item.code || "").slice(0, 16).replace(/[\^\~]/g, "");
  const barcodeStr = (item.barcode || item.code || "123456789").replace(/[\^\~]/g, "");

  return `^XA
^PW400
^LL200
^FO20,15^A0N,22,22^FD${nameStr}^FS
^FO20,42^A0N,18,18^FDSKU:${codeStr} ${mrpStr ? " " + mrpStr : ""}^FS
^FO20,68^BY2,3,38^BCN,38,Y,N,N^FD${barcodeStr}^FS
^FO20,122^A0N,22,22^FDPRICE: ${priceStr}^FS
^PQ${Math.max(1, copies)}
^XZ`;
}

/**
 * Universal PRN Generator mapping Item Master products into raw TSPL/ZPL printer commands.
 */
export function generatePRNScript(
  items: Array<{ product: Product; copies: number }>,
  options: PRNOptions
): string {
  if (!items || items.length === 0) return "";
  const generator = options.language === "ZPL" ? generateZPLScript : generateTSPLScript;
  return items
    .map(({ product, copies }) => generator(product, copies, options))
    .join("\n\n");
}
